File: apps/isomer-go/internal/sidecar/server_test.go
// HTTP contract tests for the isomer-go sidecar routes.
//
// These tests lock the public route behavior: health, bad-body handling,
// required-token failures, request forwarding into the verifier seam, and panic
// recovery.
package sidecar

import (
	"bytes"
	"context"
	"encoding/json"
	"log"
	"net/http"
	"net/http/httptest"
	"regexp"
	"strings"
	"testing"
	"time"
)

// fakeVerifier records the request data the handlers forward into the runtime
// seam and can optionally panic to exercise recovery behavior.
type fakeVerifier struct {
	vcResult *verificationResult
	vpResult *verificationResult

	vcPanic bool
	vpPanic bool

	lastVCToken string
	lastVPToken string
	lastVPAud   string
	lastVPNonce string
}

// fakeWebhook records successful webhook attempts.
type fakeWebhook struct {
	presentationCalls int
	credentialCalls   int
}

// SendPresentation satisfies the presentation webhook seam.
func (f *fakeWebhook) SendPresentation(_ context.Context, _ *verificationResult) string {
	f.presentationCalls++
	return ""
}

// SendCredential satisfies the credential webhook seam.
func (f *fakeWebhook) SendCredential(_ context.Context, _ *verificationResult) string {
	f.credentialCalls++
	return ""
}

// VerifyVC satisfies the handler seam for VC route tests.
func (f *fakeVerifier) VerifyVC(_ context.Context, token string) *verificationResult {
	if f.vcPanic {
		panic("boom")
	}
	f.lastVCToken = token
	return f.vcResult
}

// VerifyVP satisfies the handler seam for VP route tests.
func (f *fakeVerifier) VerifyVP(_ context.Context, token, audience, nonce string) *verificationResult {
	if f.vpPanic {
		panic("boom")
	}
	f.lastVPToken = token
	f.lastVPAud = audience
	f.lastVPNonce = nonce
	return f.vpResult
}

func TestHealthEndpoint(t *testing.T) {
	server := newServerWithVerifier(Config{}, &fakeVerifier{})
	response := sendRequest(t, server.handler(), http.MethodGet, "/healthz", "")

	if response.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d", response.Code)
	}

	body := decodeBody(t, response)
	if body["ok"] != true {
		t.Fatalf("expected ok health response, got %#v", body)
	}
	if body["service"] != "isomer-go" {
		t.Fatalf("expected service name, got %#v", body)
	}
}

func TestVerifyVCHandlerReturnsAcceptedOperation(t *testing.T) {
	verifier := &fakeVerifier{
		vcResult: &verificationResult{
			OK:       true,
			Kind:     "vc+jwt",
			Errors:   []string{},
			Warnings: []string{},
			Checks:   map[string]any{"jwtEnvelopeValid": true},
		},
	}
	server := newServerWithVerifier(Config{}, verifier)

	var name string
	logs := captureLogs(t, func() {
		response := sendRequest(t, server.handler(), http.MethodPost, "/verify/vc", `{"token":"abc"}`)
		if response.Code != http.StatusAccepted {
			t.Fatalf("expected 202, got %d", response.Code)
		}

		body := decodeBody(t, response)
		name, _ = body["name"].(string)
		if !regexp.MustCompile(`^verify-vc\.`).MatchString(name) {
			t.Fatalf("expected verify-vc operation name, got %#v", body)
		}
		if body["done"] != false {
			t.Fatalf("expected pending operation, got %#v", body)
		}

		listResponse := sendRequest(t, server.handler(), http.MethodGet, "/operations", "")
		if listResponse.Code != http.StatusOK {
			t.Fatalf("expected operations list 200, got %d", listResponse.Code)
		}
		listBody := decodeArrayBody(t, listResponse)
		if len(listBody) != 1 || listBody[0]["name"] != name {
			t.Fatalf("expected operation in list, got %#v", listBody)
		}

		filteredResponse := sendRequest(t, server.handler(), http.MethodGet, "/operations?type=verify-vc", "")
		if filteredResponse.Code != http.StatusOK {
			t.Fatalf("expected filtered operations list 200, got %d", filteredResponse.Code)
		}
		filteredBody := decodeArrayBody(t, filteredResponse)
		if len(filteredBody) != 1 || filteredBody[0]["name"] != name {
			t.Fatalf("expected filtered operation in list, got %#v", filteredBody)
		}

		completed := waitForOperation(t, server.handler(), name)
		if verifier.lastVCToken != "abc" {
			t.Fatalf("expected token to reach verifier, got %q", verifier.lastVCToken)
		}
		metadata, _ := completed["metadata"].(map[string]any)
		if metadata["state"] != "completed" {
			t.Fatalf("expected completed operation, got %#v", completed)
		}
		responseBody, _ := completed["response"].(map[string]any)
		if responseBody["kind"] != "vc+jwt" {
			t.Fatalf("expected VC result in operation response, got %#v", completed)
		}
	})
	received := findLog(t, logs, "verification.received", map[string]any{"operationName": name})
	if received["token"] != "abc" || received["tokenLength"] != float64(3) || received["tokenSha256"] != "ba7816bf8f01cfea" {
		t.Fatalf("unexpected received log: %#v", received)
	}
	resultLog := findLog(t, logs, "verification.result", map[string]any{"operationName": name})
	if resultLog["ok"] != true || resultLog["kind"] != "vc+jwt" {
		t.Fatalf("unexpected result log: %#v", resultLog)
	}
	skipped := findLog(t, logs, "webhook.skipped", map[string]any{"artifactKind": "vc+jwt"})
	if skipped["reason"] != "no_webhook_url" {
		t.Fatalf("unexpected webhook skipped log: %#v", skipped)
	}
}

func TestVerifyVCHandlerStoresFailedOperation(t *testing.T) {
	server := newServerWithVerifier(Config{}, &fakeVerifier{vcPanic: true})

	var name string
	logs := captureLogs(t, func() {
		response := sendRequest(t, server.handler(), http.MethodPost, "/verify/vc", `{"token":"abc"}`)
		if response.Code != http.StatusAccepted {
			t.Fatalf("expected 202, got %d", response.Code)
		}
		body := decodeBody(t, response)
		name, _ = body["name"].(string)

		failed := waitForOperation(t, server.handler(), name)
		metadata, _ := failed["metadata"].(map[string]any)
		if metadata["state"] != "failed" {
			t.Fatalf("expected failed operation, got %#v", failed)
		}
		errorBody, _ := failed["error"].(map[string]any)
		if errorBody["code"] != float64(http.StatusInternalServerError) || !strings.Contains(errorBody["message"].(string), "sidecar panic") {
			t.Fatalf("expected panic operation error, got %#v", failed)
		}
	})
	resultLog := findLog(t, logs, "verification.result", map[string]any{"operationName": name})
	if resultLog["ok"] != false || !strings.Contains(resultLog["errors"].([]any)[0].(string), "sidecar panic") {
		t.Fatalf("unexpected failure log: %#v", resultLog)
	}
}

func TestVerifyVPHandlerPassesAudienceAndNonce(t *testing.T) {
	verifier := &fakeVerifier{
		vpResult: &verificationResult{
			OK:       true,
			Kind:     "vp+jwt",
			Errors:   []string{},
			Warnings: []string{},
			Checks:   map[string]any{"jwtEnvelopeValid": true},
		},
	}
	server := newServerWithVerifier(Config{}, verifier)

	logs := captureLogs(t, func() {
		response := sendRequest(t, server.handler(), http.MethodPost, "/verify/vp", `{"token":"abc","audience":"aud","nonce":"n-1"}`)
		if response.Code != http.StatusOK {
			t.Fatalf("expected 200, got %d", response.Code)
		}
		if verifier.lastVPToken != "abc" || verifier.lastVPAud != "aud" || verifier.lastVPNonce != "n-1" {
			t.Fatalf("unexpected verifier inputs: token=%q audience=%q nonce=%q", verifier.lastVPToken, verifier.lastVPAud, verifier.lastVPNonce)
		}
	})
	received := findLog(t, logs, "verification.received", map[string]any{"artifactKind": "vp+jwt"})
	if received["route"] != "/verify/vp" || received["token"] != "abc" {
		t.Fatalf("unexpected received log: %#v", received)
	}
	resultLog := findLog(t, logs, "verification.result", map[string]any{"artifactKind": "vp+jwt"})
	if resultLog["ok"] != true || resultLog["kind"] != "vp+jwt" {
		t.Fatalf("unexpected result log: %#v", resultLog)
	}
}

func TestVerifyVPHandlerSendsWebhookAfterSuccess(t *testing.T) {
	verifier := &fakeVerifier{
		vpResult: &verificationResult{
			OK:       true,
			Kind:     "vp+jwt",
			Errors:   []string{},
			Warnings: []string{},
			Checks:   map[string]any{"jwtEnvelopeValid": true},
		},
	}
	webhook := &fakeWebhook{}
	server := newServerWithVerifierAndWebhook(Config{WebhookURL: "http://dashboard.test/webhooks/presentations"}, verifier, webhook)

	response := sendRequest(t, server.handler(), http.MethodPost, "/verify/vp", `{"token":"abc"}`)
	if response.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d", response.Code)
	}
	if webhook.presentationCalls != 1 {
		t.Fatalf("expected one presentation webhook call, got %d", webhook.presentationCalls)
	}
}

func TestVerifyVCHandlerSendsWebhookAfterSuccess(t *testing.T) {
	verifier := &fakeVerifier{
		vcResult: &verificationResult{
			OK:       true,
			Kind:     "vc+jwt",
			Errors:   []string{},
			Warnings: []string{},
			Checks:   map[string]any{"jwtEnvelopeValid": true},
		},
	}
	webhook := &fakeWebhook{}
	server := newServerWithVerifierAndWebhook(Config{WebhookURL: "http://dashboard.test/webhooks/presentations"}, verifier, webhook)

	response := sendRequest(t, server.handler(), http.MethodPost, "/verify/vc", `{"token":"abc"}`)
	if response.Code != http.StatusAccepted {
		t.Fatalf("expected 202, got %d", response.Code)
	}
	body := decodeBody(t, response)
	name, _ := body["name"].(string)
	completed := waitForOperation(t, server.handler(), name)
	metadata, _ := completed["metadata"].(map[string]any)
	if metadata["state"] != "completed" {
		t.Fatalf("expected completed operation, got %#v", completed)
	}
	if webhook.credentialCalls != 1 {
		t.Fatalf("expected one credential webhook call, got %d", webhook.credentialCalls)
	}
}

func TestVerifyHandlersRejectBadJSON(t *testing.T) {
	server := newServerWithVerifier(Config{}, &fakeVerifier{})

	for _, path := range []string{"/verify/vc", "/verify/vp"} {
		response := sendRequest(t, server.handler(), http.MethodPost, path, `{"token":`)
		if response.Code != http.StatusBadRequest {
			t.Fatalf("%s: expected 400, got %d", path, response.Code)
		}
		body := decodeBody(t, response)
		if body["ok"] != false {
			t.Fatalf("%s: expected ok=false, got %#v", path, body)
		}
	}
}

func TestVerifyHandlersRequireToken(t *testing.T) {
	server := newServerWithVerifier(Config{}, &fakeVerifier{})

	for _, path := range []string{"/verify/vc", "/verify/vp"} {
		response := sendRequest(t, server.handler(), http.MethodPost, path, `{}`)
		if response.Code != http.StatusBadRequest {
			t.Fatalf("%s: expected 400, got %d", path, response.Code)
		}
		body := decodeBody(t, response)
		if body["error"] != "verification request requires token" {
			t.Fatalf("%s: unexpected body %#v", path, body)
		}
	}
}

func TestRecoverHandlerReturnsJSONOnVPVerifierPanic(t *testing.T) {
	server := newServerWithVerifier(Config{}, &fakeVerifier{vpPanic: true})

	response := sendRequest(t, server.handler(), http.MethodPost, "/verify/vp", `{"token":"abc"}`)
	if response.Code != http.StatusInternalServerError {
		t.Fatalf("expected 500, got %d", response.Code)
	}
	body := decodeBody(t, response)
	if body["kind"] != "sidecar" {
		t.Fatalf("expected sidecar panic body, got %#v", body)
	}
	errors, _ := body["errors"].([]any)
	if len(errors) != 1 || !strings.Contains(errors[0].(string), "sidecar panic") {
		t.Fatalf("expected panic error, got %#v", body)
	}
}

// sendRequest drives one route through the in-memory HTTP handler under test.
func sendRequest(t *testing.T, handler http.Handler, method, path, body string) *httptest.ResponseRecorder {
	t.Helper()
	request := httptest.NewRequest(method, path, strings.NewReader(body))
	if method == http.MethodPost {
		request.Header.Set("Content-Type", "application/json")
	}

	response := httptest.NewRecorder()
	handler.ServeHTTP(response, request)
	return response
}

// decodeBody decodes one JSON response body into a plain map for test
// assertions.
func decodeBody(t *testing.T, response *httptest.ResponseRecorder) map[string]any {
	t.Helper()
	var body map[string]any
	if err := json.Unmarshal(response.Body.Bytes(), &body); err != nil {
		t.Fatal(err)
	}
	return body
}

func decodeArrayBody(t *testing.T, response *httptest.ResponseRecorder) []map[string]any {
	t.Helper()
	var body []map[string]any
	if err := json.Unmarshal(response.Body.Bytes(), &body); err != nil {
		t.Fatal(err)
	}
	return body
}

func waitForOperation(t *testing.T, handler http.Handler, name string) map[string]any {
	t.Helper()
	for attempt := 0; attempt < 40; attempt++ {
		response := sendRequest(t, handler, http.MethodGet, "/operations/"+name, "")
		if response.Code != http.StatusOK {
			t.Fatalf("expected operation lookup 200, got %d", response.Code)
		}
		body := decodeBody(t, response)
		if body["done"] == true {
			return body
		}
		time.Sleep(5 * time.Millisecond)
	}
	t.Fatalf("operation %s did not complete", name)
	return nil
}

func captureLogs(t *testing.T, callback func()) []map[string]any {
	t.Helper()
	var buffer bytes.Buffer
	originalWriter := log.Writer()
	originalFlags := log.Flags()
	log.SetOutput(&buffer)
	log.SetFlags(0)
	defer func() {
		log.SetOutput(originalWriter)
		log.SetFlags(originalFlags)
	}()

	callback()

	lines := strings.Split(strings.TrimSpace(buffer.String()), "\n")
	events := []map[string]any{}
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		start := strings.Index(line, "{")
		if start < 0 {
			continue
		}
		var event map[string]any
		if err := json.Unmarshal([]byte(line[start:]), &event); err != nil {
			t.Fatalf("invalid log line %q: %v", line, err)
		}
		events = append(events, event)
	}
	return events
}

func findLog(t *testing.T, logs []map[string]any, event string, fields map[string]any) map[string]any {
	t.Helper()
	for _, item := range logs {
		if item["event"] != event {
			continue
		}
		matched := true
		for key, value := range fields {
			if item[key] != value {
				matched = false
				break
			}
		}
		if matched {
			return item
		}
	}
	t.Fatalf("missing log %q with %#v in %#v", event, fields, logs)
	return nil
}

File: apps/isomer-node/src/server.test.ts
/**
 * HTTP contract tests for the isomer-node sidecar routes.
 *
 * These tests intentionally lock the public route behavior: health, bad-body
 * handling, required-token failures, and request forwarding into the verifier
 * seam.
 */
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import test from "node:test";
import { createApp, type RequestVerifier } from "./server.js";
import type { SidecarConfig, VcVerificationResult, VpVerificationResult } from "./types.js";

// Shared sidecar config used across route contract tests.
const CONFIG: SidecarConfig = {
  host: "127.0.0.1",
  port: 8787,
  resolverUrl: "http://resolver.test/1.0/identifiers",
  resourceRoot: "/tmp/resources",
  verifierId: "isomer-node-test"
};

test("health route reports service readiness", async () => {
  const app = createApp(CONFIG, undefined as never, createStubVerifier({}));

  const response = await app.fetch(new Request("http://localhost/healthz"));

  assert.equal(response.status, 200);
  assert.deepEqual(await response.json(), { ok: true, service: "isomer-node" });
});

test("vc route rejects requests without token", async () => {
  const app = createApp(CONFIG, undefined as never, createStubVerifier({}));

  const response = await app.fetch(new Request("http://localhost/verify/vc", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({})
  }));

  assert.equal(response.status, 400);
  assert.deepEqual(await response.json(), {
    ok: false,
    error: "verification request requires token"
  });
});

test("vc route treats malformed json like an empty request", async () => {
  const app = createApp(CONFIG, undefined as never, createStubVerifier({}));

  const response = await app.fetch(new Request("http://localhost/verify/vc", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: "{not-json"
  }));

  assert.equal(response.status, 400);
  assert.deepEqual(await response.json(), {
    ok: false,
    error: "verification request requires token"
  });
});

test("vc route returns 202 and stores a pending operation", async () => {
  const result: VcVerificationResult = {
    ok: true,
    kind: "vc+jwt",
    errors: [],
    warnings: [],
    payload: { id: "urn:example:vc" },
    checks: {
      jwtEnvelopeValid: true,
      signatureValid: true,
      dataIntegrityProofValid: true,
      statusActive: true
    }
  };
  let receivedToken: string | undefined;

  const app = createApp(CONFIG, undefined as never, createStubVerifier({
    verifyVc: function* (token) {
      receivedToken = token;
      return result;
    }
  }));

  let operationName = "";
  const logs = await captureConsoleLogs(async () => {
    const response = await app.fetch(new Request("http://localhost/verify/vc", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ token: "vc-token" })
    }));

    assert.equal(response.status, 202);
    const operation = await response.json();
    operationName = operation.name;
    assert.match(operation.name, /^verify-vc\./);
    assert.equal(operation.done, false);
    assert.equal(operation.metadata.state, "pending");

    const listResponse = await app.fetch(new Request("http://localhost/operations"));
    assert.equal(listResponse.status, 200);
    const listBody = await listResponse.json() as Array<{ name: string }>;
    assert.deepEqual(listBody.map((item) => item.name), [operation.name]);

    const filteredResponse = await app.fetch(new Request("http://localhost/operations?type=verify-vc"));
    assert.equal(filteredResponse.status, 200);
    const filteredBody = await filteredResponse.json() as Array<{ name: string }>;
    assert.deepEqual(filteredBody.map((item) => item.name), [operation.name]);

    const completed = await waitForOperation(app, operation.name);
    assert.equal(receivedToken, "vc-token");
    assert.equal(completed.done, true);
    assert.equal(completed.metadata.state, "completed");
    assert.deepEqual(completed.response, result);
  });

  const received = findLog(logs, "verification.received", { operationName });
  assert.equal(received.verifier, "isomer-node-test");
  assert.equal(received.route, "/verify/vc");
  assert.equal(received.artifactKind, "vc+jwt");
  assert.equal(received.token, "vc-token");
  assert.equal(received.tokenLength, "vc-token".length);
  assert.equal(received.tokenSha256, tokenHash("vc-token"));
  const resultLog = findLog(logs, "verification.result", { operationName });
  assert.equal(resultLog.ok, true);
  assert.equal(resultLog.kind, "vc+jwt");
  const skipped = findLog(logs, "webhook.skipped", { artifactKind: "vc+jwt" });
  assert.equal(skipped.reason, "no_webhook_url");
});

test("vc route records failed background verification operations", async () => {
  const app = createApp(CONFIG, undefined as never, createStubVerifier({
    verifyVc: function* () {
      throw new Error("verification crashed");
    }
  }));

  let operationName = "";
  const logs = await captureConsoleLogs(async () => {
    const response = await app.fetch(new Request("http://localhost/verify/vc", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ token: "vc-token" })
    }));

    assert.equal(response.status, 202);
    const operation = await response.json();
    operationName = operation.name;
    const failed = await waitForOperation(app, operation.name);
    assert.equal(failed.done, true);
    assert.equal(failed.metadata.state, "failed");
    assert.equal(failed.error.code, 500);
    assert.equal(failed.error.message, "verification crashed");
  });

  const resultLog = findLog(logs, "verification.result", { operationName });
  assert.equal(resultLog.ok, false);
  assert.deepEqual(resultLog.errors, ["verification crashed"]);
  assert.deepEqual(resultLog.error, { code: 500, message: "verification crashed" });
});

test("vp route forwards audience and nonce", async () => {
  const result: VpVerificationResult = {
    ok: true,
    kind: "vp+jwt",
    errors: [],
    warnings: [],
    payload: { holder: "did:webs:holder" },
    checks: {
      jwtEnvelopeValid: true,
      signatureValid: true,
      embeddedCredentialsVerified: 1
    },
    nested: [{
      ok: true,
      kind: "vc+jwt",
      errors: [],
      warnings: [],
      payload: { id: "urn:example:vc" },
      checks: {
        jwtEnvelopeValid: true,
        signatureValid: true,
        dataIntegrityProofValid: true,
        statusActive: true
      }
    }]
  };
  let received: { token?: string; audience?: string; nonce?: string } = {};

  const app = createApp(CONFIG, undefined as never, createStubVerifier({
    verifyVp: function* (token, options) {
      received = {
        token,
        audience: options.audience,
        nonce: options.nonce
      };
      return result;
    }
  }));

  const logs = await captureConsoleLogs(async () => {
    const response = await app.fetch(new Request("http://localhost/verify/vp", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        token: "vp-token",
        audience: "https://verifier.example",
        nonce: "expected-nonce"
      })
    }));

    assert.deepEqual(received, {
      token: "vp-token",
      audience: "https://verifier.example",
      nonce: "expected-nonce"
    });
    assert.equal(response.status, 200);
    assert.deepEqual(await response.json(), result);
  });

  const receivedLog = findLog(logs, "verification.received", { artifactKind: "vp+jwt" });
  assert.equal(receivedLog.route, "/verify/vp");
  assert.equal(receivedLog.token, "vp-token");
  const resultLog = findLog(logs, "verification.result", { artifactKind: "vp+jwt" });
  assert.equal(resultLog.ok, true);
  assert.equal(resultLog.kind, "vp+jwt");
});

test("vp route sends webhook only after successful presentation verification", async () => {
  const result: VpVerificationResult = {
    ok: true,
    kind: "vp+jwt",
    errors: [],
    warnings: [],
    payload: { holder: "did:webs:holder" },
    checks: {
      jwtEnvelopeValid: true,
      signatureValid: true,
      embeddedCredentialsVerified: 0
    },
    nested: []
  };
  let webhookCalls = 0;

  const app = createApp(
    { ...CONFIG, webhookUrl: "http://dashboard.test/webhooks/presentations" },
    undefined as never,
    createStubVerifier({ verifyVp: function* () { return result; } }),
    {
      sendPresentation: async () => { webhookCalls += 1; return null; },
      sendCredential: async () => null
    }
  );

  const logs = await captureConsoleLogs(async () => {
    const response = await app.fetch(new Request("http://localhost/verify/vp", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ token: "vp-token" })
    }));

    assert.equal(response.status, 200);
    assert.equal(webhookCalls, 1);
    assert.deepEqual(await response.json(), result);
  });
  assert.equal(findLog(logs, "verification.received", { artifactKind: "vp+jwt" }).token, "vp-token");
  assert.equal(findLog(logs, "verification.result", { artifactKind: "vp+jwt" }).ok, true);
});

test("vc route sends webhook after successful credential verification", async () => {
  let webhookCalls = 0;
  const app = createApp(
    { ...CONFIG, webhookUrl: "http://dashboard.test/webhooks/presentations" },
    undefined as never,
    createStubVerifier({}),
    {
      sendPresentation: async () => null,
      sendCredential: async () => { webhookCalls += 1; return null; }
    }
  );

  let operationName = "";
  const logs = await captureConsoleLogs(async () => {
    const response = await app.fetch(new Request("http://localhost/verify/vc", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ token: "vc-token" })
    }));

    assert.equal(response.status, 202);
    const operation = await response.json();
    operationName = operation.name;
    const completed = await waitForOperation(app, operation.name);
    assert.equal(completed.metadata.state, "completed");
    assert.equal(webhookCalls, 1);
  });
  assert.equal(findLog(logs, "verification.received", { operationName }).token, "vc-token");
  assert.equal(findLog(logs, "verification.result", { operationName }).ok, true);
});

/**
 * Create a stub request verifier for route tests.
 *
 * The helper defaults to successful no-op results so each test overrides only
 * the branch it cares about.
 */
function createStubVerifier(overrides: {
  verifyVc?: RequestVerifier["verifyVc"];
  verifyVp?: RequestVerifier["verifyVp"];
}): RequestVerifier {
  return {
    verifyVc: overrides.verifyVc ?? (function* () {
        return {
          ok: true,
          kind: "vc+jwt" as const,
          errors: [],
          warnings: [],
          payload: null,
          checks: {
            jwtEnvelopeValid: true,
            signatureValid: true,
            dataIntegrityProofValid: true,
            statusActive: true
          }
        };
      }),
    verifyVp: overrides.verifyVp ?? (
      function* () {
        return {
          ok: true,
          kind: "vp+jwt" as const,
          errors: [],
          warnings: [],
          payload: null,
          checks: {
            jwtEnvelopeValid: true,
            signatureValid: true,
            embeddedCredentialsVerified: 0
          },
          nested: []
        };
      })
  };
}

async function waitForOperation(
  app: ReturnType<typeof createApp>,
  name: string
): Promise<any> {
  for (let attempt = 0; attempt < 20; attempt += 1) {
    const response = await app.fetch(new Request(`http://localhost/operations/${name}`));
    assert.equal(response.status, 200);
    const operation = await response.json();
    if (operation.done === true) {
      return operation;
    }
    await new Promise((resolve) => setTimeout(resolve, 5));
  }
  throw new Error(`operation ${name} did not complete`);
}

async function captureConsoleLogs(callback: () => Promise<void>): Promise<Array<Record<string, any>>> {
  const original = console.log;
  const lines: string[] = [];
  console.log = (message?: unknown) => {
    lines.push(String(message));
  };
  try {
    await callback();
  } finally {
    console.log = original;
  }
  return lines.map((line) => JSON.parse(line) as Record<string, any>);
}

function findLog(
  logs: Array<Record<string, any>>,
  event: string,
  fields: Record<string, unknown> = {}
): Record<string, any> {
  const match = logs.find((item) => (
    item.event === event &&
    Object.entries(fields).every(([key, value]) => item[key] === value)
  ));
  if (match === undefined) {
    throw new Error(`missing log ${event} with ${JSON.stringify(fields)} in ${JSON.stringify(logs)}`);
  }
  return match;
}

function tokenHash(token: string): string {
  return createHash("sha256").update(token, "utf8").digest("hex").slice(0, 16);
}
